src/Common/src/System/Collections/Generic/EnumerableHelpers.cs

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace System.Collections.Generic
+{
+    /// <summary>Internal helper functions for working with enumerables.</summary>
+    internal static class EnumerableHelpers
+    {
+        /// <summary>Converts an enumerable to an array using the same logic as does List{T}.</summary>
+        /// <param name="count">The number of items stored in the resulting array, 0-indexed.</param>
+        /// <returns>
+        /// The resulting array.  The length of the array may be greater than <paramref name="count"/>,
+        /// which is the actual number of elements in the array.
+        /// </returns>
+        internal static T[] ToArray<T>(IEnumerable<T> source, out int count)
+        {
+            T[] arr;
+            int found = 0;
+
+            ICollection<T> ic = source as ICollection<T>;
+            if (ic != null)
+            {
+                found = ic.Count;
+                if (found == 0)
+                {
+                    arr = Array.Empty<T>();
+                }
+                else
+                {
+                    // Allocate an array of the desired size, then copy the elements into it. Note that this has the same 
+                    // issue regarding concurrency as other existing collections like List<T>. If the collection size 
+                    // concurrently changes between the array allocation and the CopyTo, we could end up either getting an 
+                    // exception from overrunning the array (if the size went up) or we could end up not filling as many 
+                    // items as 'count' suggests (if the size went down).  This is only an issue for concurrent collections 
+                    // that implement ICollection<T>, which as of .NET 4.6 is just ConcurrentDictionary<TKey, TValue>.
+                    arr = new T[found];
+                    ic.CopyTo(arr, 0);
+                }
+            }
+            else
+            {
+                arr = Array.Empty<T>();
+                foreach (var item in source)
+                {
+                    if (found == arr.Length)
+                    {
+                        const int DefaultCapacity = 4;
+                        const int MaxArrayLength = 0x7FEFFFFF; // from Array.MaxArrayLength
+
+                        int newLength = found == 0 ? DefaultCapacity : found * 2;
+                        if ((uint)newLength > MaxArrayLength) newLength = MaxArrayLength;
+                        if (newLength < found + 1) newLength = found + 1;
+
+                        arr = ArrayT<T>.Resize(arr, newLength, found);
```

Comment:
The corefx codebase is currently inconsistent about whether it uses ```Array.Resize``` or ```ArrayT<T>.Resize```, the latter of which was added for performance reasons with .NET Native.  I chose to use the latter for now, as it's what's used elsewhere in System.Collections.  We do separately need to rationalize its use across the codebase, and ideally get whatever additional methods or improvements are necessary into the real Array rather than relying on a separate method like this.